Support describe cluster in kafka client

michael-redpanda opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

With Topic ID support added, the metadata API handler was updated to v12.  v11 dropped support for the cluster_authorized_operations and moved that to the new DescribeCluster API.  This PR now uses that API to fetch that information and broker information and only relies on the Metadata API for topic information.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* None

michael-redpanda committed:
k/client: Make the fetch for API version generic

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/p: Use a named type for cluster authorized ops

Signed-off-by: Michael Boquard <michael@redpanda.com>

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/72223-->
<!--5f357eb1-7bb4-4a69-a656-84d7fa19caf2-->
<!--88d1fa93-6386-4f3e-b345-b7969f420fc6-->
<!--3d5465b0-242b-49ff-aa0d-fe8e84c14828-->
<!--7e536d80-2175-4885-a843-cf1c7927ac1b-->
<!--7d362179-1f0b-4882-8471-ef9b2b45907c-->
<!--82c4bd67-f275-403e-a386-bc439604a10f-->
<!--8e850e4f-a486-4fb4-9ac2-ba7c644b8add-->
<!--title-->
## Retry command for [Build#72223](https://buildkite.com/redpanda/redpanda/builds/72223)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"adjacent_merge","enable_failures":false,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"sliding_window","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"chunked_sliding_window","enable_failures":false,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"sliding_window","enable_failures":false,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"adjacent_merge","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"adjacent_merge","enable_failures":false,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"sliding_window","enable_failures":false,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"chunked_sliding_window","enable_failures":false,"mixed_versions":true,"with_iceberg":false}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--838e837c-5927-41f9-9292-545b7d6c8102-->
<!--TEST RESULTS-->
<!--80790f05-d37c-49b4-889a-e64336c11b19-->
<!--TEST RESULTS-->
<!--3eba0417-9b7f-4513-bbf0-8185fd9f98c3-->
<!--TEST RESULTS-->
<!--ab6b3f77-de04-4dd7-b49e-c618a2fad3c7-->
<!--TEST RESULTS-->
<!--e1294244-d4ad-4baf-9b04-6008320b7592-->
<!--TEST RESULTS-->
<!--62278ce8-ade5-4457-903a-77b6fdb9c9fc-->
<!--TEST RESULTS-->
<!--d3359818-b967-4cc4-9a1d-40bc24c1b8ed-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72223">build#72223</a></summary>


| test_class               | test_method                         | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionBalancerTest    | test_recovery_mode_rebalance_finish | null                                                                                                                                            | integration | https://buildkite.com/redpanda/redpanda/builds/72223#0199444f-4d62-4159-b217-fae3c1190f8d | FLAKY         | 12/21    | upstream reliability is '94.89795918367348'. current run reliability is '57.14285714285714'. drift is 37.7551 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations                | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/72223#0199444f-4d5c-40a9-bcf4-b238b1da650d | FAIL          | 0/1      |                                                                                                                                                                        |
| RandomNodeOperationsTest | test_node_operations                | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/72223#0199444f-4d5d-42c2-a2f3-aa2787c413ee | FAIL          | 0/1      |                                                                                                                                                                        |
| RandomNodeOperationsTest | test_node_operations                | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72223#0199444f-4d5e-4089-b076-907306c5c4d4 | FAIL          | 0/1      |                                                                                                                                                                        |
| RandomNodeOperationsTest | test_node_operations                | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72223#0199444f-4d5f-4486-9c47-279ba22832d0 | FAIL          | 0/1      |                                                                                                                                                                        |
| RandomNodeOperationsTest | test_node_operations                | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/72223#0199444f-4d60-43ce-9e25-ab3ff10cea02 | FAIL          | 0/1      |                                                                                                                                                                        |
| RandomNodeOperationsTest | test_node_operations                | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/72223#0199444f-4d61-4f5f-a076-7e331d7ff472 | FAIL          | 0/1      |                                                                                                                                                                        |
| RandomNodeOperationsTest | test_node_operations                | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}          | integration | https://buildkite.com/redpanda/redpanda/builds/72223#0199444f-4d61-4f5f-a076-7e331d7ff472 | FAIL          | 0/1      |                                                                                                                                                                        |
| RandomNodeOperationsTest | test_node_operations                | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}          | integration | https://buildkite.com/redpanda/redpanda/builds/72223#0199444f-4d5b-422a-aa69-36252ea52cf6 | FAIL          | 0/1      |                                                                                                                                                                        |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72256">build#72256</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:-------------------------|:---------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| AuditLogTestOauth        | test_kafka_oauth     | {"audit_transport_mode": "kclient", "authz_match": "acl"}                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/72256#01994d5c-45f4-44ea-a455-ed72cb956f6f | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72256#01994d5c-45fb-48d3-8a06-87d178af8024 | FAIL          | 0/1      |                                                                                                                                                           |
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}  | integration | https://buildkite.com/redpanda/redpanda/builds/72256#01994d5c-45ff-4fd9-9f3b-20468d968345 | FAIL          | 0/1      |                                                                                                                                                           |
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}  | integration | https://buildkite.com/redpanda/redpanda/builds/72256#01994d5c-45f3-44bb-974c-057cbfa43fff | FAIL          | 0/1      |                                                                                                                                                           |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72308">build#72308</a></summary>


| test_class                    | test_method                    | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                      |
|:------------------------------|:-------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_replication_with_failures | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/72308#01995050-df60-40f9-8a4d-64c9efe95931 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest      | test_node_operations           | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false}        | integration | https://buildkite.com/redpanda/redpanda/builds/72308#0199504d-f29d-4308-9073-9cd4c82709cb | FAIL          | 0/1      |                                                                                                                                                             |
| RandomNodeOperationsTest      | test_node_operations           | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/72308#0199504d-f2a0-4775-9f87-c9c9badeb0c9 | FLAKY         | 10/21    | upstream reliability is '100.0'. current run reliability is '47.61904761904761'. drift is 52.38095 and the allowed drift is set to 50. The test should FAIL |
| RandomNodeOperationsTest      | test_node_operations           | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/72308#0199504d-f2a1-47a2-abe8-8213600050fd | FAIL          | 0/1      |                                                                                                                                                             |
| RandomNodeOperationsTest      | test_node_operations           | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72308#0199504d-f2a2-4e23-b91a-cf062107d548 | FAIL          | 0/1      |                                                                                                                                                             |
| RandomNodeOperationsTest      | test_node_operations           | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/72308#0199504d-f29c-4799-8dd9-aa278a1429b2 | FAIL          | 0/1      |                                                                                                                                                             |
| RandomNodeOperationsTest      | test_node_operations           | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/72308#0199504d-f29d-4c72-9d03-56b4533c5d2e | FAIL          | 0/1      |                                                                                                                                                             |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72331">build#72331</a></summary>


| test_class               | test_method                           | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:-------------------------|:--------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MasterTestSuite          | test_chunk_dl_with_random_http_errors |                                                                                                                                                 | unit        | https://buildkite.com/redpanda/redpanda/builds/72331#01995220-ed96-44eb-967f-6edd0cb6a9fe | FAIL          | 0/1      |                                                                                                                                                                       |
| PartitionBalancerTest    | test_recovery_mode_rebalance_finish   | null                                                                                                                                            | integration | https://buildkite.com/redpanda/redpanda/builds/72331#01995284-dc6c-46f4-a4ac-10d6ef75a837 | FLAKY         | 16/21    | upstream reliability is '83.9080459770115'. current run reliability is '76.19047619047619'. drift is 7.71757 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations                  | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72331#01995284-dc69-445f-a2d1-9bceaf48df62 | FAIL          | 0/1      |                                                                                                                                                                       |
| RandomNodeOperationsTest | test_node_operations                  | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72331#01995284-dc6a-4a87-abdd-5023de668f1d | FAIL          | 0/1      |                                                                                                                                                                       |
| RandomNodeOperationsTest | test_node_operations                  | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/72331#01995284-dc6b-4fce-b77b-64d48d089500 | FAIL          | 0/1      |                                                                                                                                                                       |
| RandomNodeOperationsTest | test_node_operations                  | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/72331#01995284-dc6c-46ef-8fda-a1bdc401496c | FAIL          | 0/1      |                                                                                                                                                                       |
| RandomNodeOperationsTest | test_node_operations                  | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}  | integration | https://buildkite.com/redpanda/redpanda/builds/72331#01995284-dc6c-46f4-a4ac-10d6ef75a837 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                                |
| RandomNodeOperationsTest | test_node_operations                  | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}          | integration | https://buildkite.com/redpanda/redpanda/builds/72331#01995284-dc66-466a-ab5c-95128e85b5d2 | FAIL          | 0/1      |                                                                                                                                                                       |
| RandomNodeOperationsTest | test_node_operations                  | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}          | integration | https://buildkite.com/redpanda/redpanda/builds/72331#01995284-dc67-4f24-9f10-c3c83f800ff2 | FAIL          | 0/1      |                                                                                                                                                                       |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72406">build#72406</a></summary>


| test_class                           | test_method                | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------------------|:---------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync | {"with_failures": true}                                                                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/72406#01995573-58c7-450a-8b61-3fa973fcf89b | FLAKY         | 18/21    | upstream reliability is '94.11764705882352'. current run reliability is '85.71428571428571'. drift is 8.40336 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest             | test_node_operations       | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/72406#01995573-58ca-4ceb-abf6-afd3d3e788d1 | FAIL          | 0/1      |                                                                                                                                                                        |
| RandomNodeOperationsTest             | test_node_operations       | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/72406#01995573-58cb-478b-923e-6524f23ce996 | FAIL          | 0/1      |                                                                                                                                                                        |
| RandomNodeOperationsTest             | test_node_operations       | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72406#01995573-58cc-4f6a-a229-b0e858eee67e | FAIL          | 0/1      |                                                                                                                                                                        |
| RandomNodeOperationsTest             | test_node_operations       | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}          | integration | https://buildkite.com/redpanda/redpanda/builds/72406#01995573-58c9-4eef-94b4-e6dd7cec1cbd | FAIL          | 0/1      |                                                                                                                                                                        |
| RandomNodeOperationsTest             | test_node_operations       | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}          | integration | https://buildkite.com/redpanda/redpanda/builds/72406#01995573-58ca-4011-a310-f02c944a30d4 | FAIL          | 0/1      |                                                                                                                                                                        |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72443">build#72443</a></summary>


| test_class                 | test_method                | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:---------------------------|:---------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest   | test_node_operations       | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72443#01995800-6932-4dcc-aa49-b378853297a7 | FAIL          | 0/1      |                                                                                                                                                                         |
| RandomNodeOperationsTest   | test_node_operations       | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/72443#01995800-692b-4101-85cf-b7220248fb86 | FAIL          | 0/1      |                                                                                                                                                                         |
| RandomNodeOperationsTest   | test_node_operations       | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}          | integration | https://buildkite.com/redpanda/redpanda/builds/72443#01995800-692a-4ab5-8787-62b9f54c2d6d | FAIL          | 0/1      |                                                                                                                                                                         |
| RandomNodeOperationsTest   | test_node_operations       | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}          | integration | https://buildkite.com/redpanda/redpanda/builds/72443#01995800-692b-4101-85cf-b7220248fb86 | FAIL          | 0/1      |                                                                                                                                                                         |
| SchemaRegistryAutoAuthTest | test_restarts              | {"move_controller_leader": true}                                                                                                                | integration | https://buildkite.com/redpanda/redpanda/builds/72443#01995800-692f-4f71-9505-5a33c28eca5d | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| SimpleEndToEndTest         | test_consumer_interruption | null                                                                                                                                            | integration | https://buildkite.com/redpanda/redpanda/builds/72443#01995800-c863-463b-8976-fa8b4d1331a8 | FLAKY         | 17/21    | upstream reliability is '98.74476987447699'. current run reliability is '80.95238095238095'. drift is 17.79239 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72535">build#72535</a></summary>


| test_class                   | test_method        | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                  |
|:-----------------------------|:-------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionMovementUpgradeTest | test_basic_upgrade | null             | integration | https://buildkite.com/redpanda/redpanda/builds/72535#01995cdf-41fd-4783-abeb-64d6ae4778ce | FLAKY         | 20/21    | upstream reliability is '99.76958525345621'. current run reliability is '95.23809523809523'. drift is 4.53149 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementUpgradeTest&test_method=test_basic_upgrade |
</details>
<!--/body-->

mmaslankaprv commented on src/v/kafka/client/cluster.h:
```diff
@@ -97,6 +97,7 @@ class cluster {
               return std::get<Ret>(std::move(response));
           });
     }
+    using topic_request_list_t = std::optional<chunked_vector<model::topic>>;
```
nit: i do not think we need this type alias, for me it reduces readability

mmaslankaprv commented on src/v/kafka/client/cluster.cc:
```diff
@@ -194,6 +196,26 @@ ss::future<> cluster::dispatch_metadata_request() {
+              };
+          });
+        return topics;
+    }();
```
why this is a lambda, which is immediately called ? can you just make it plain code which is part of the method ?

michael-redpanda commented:
Force push:

* Addressed PR comments

BenPope commented on src/v/kafka/client/cluster.cc:
```diff
@@ -203,7 +225,9 @@ ss::future<> cluster::dispatch_metadata_request() {
-          metadata_request{.data{.include_topic_authorized_operations = true}},
+          metadata_request{.data{
+            .topics = std::move(metadata_topics_request),
+            .include_topic_authorized_operations = true}},
```
I know there were a lot of changes in this area, is all of the moving safe under any retries now?

michael-redpanda replied:
thanks for bringing this up - I actually think the caller is responsible for the retry

BenPope commented on src/v/kafka/client/cluster.cc:
```diff
@@ -173,46 +174,106 @@ ss::future<> cluster::initialize_metadata_with_seed() {
 
     if (_brokers.empty()) {
-        // If there are no brokers, connect to one of the seeds
         co_await initialize_metadata_with_seed();
```
I agree, no need to declare this so high in the function and assign it later, but it's not part of this PR.

michael-redpanda replied:
`_brokers.any()` will throw if there are no brokers, so this is handled

BenPope replied:
I only agreed with the first bit: "The variable broker is declared but not initialized."

BenPope commented on src/v/kafka/client/brokers.cc:
```diff
@@ -58,13 +58,13 @@ ss::future<> brokers::do_erase(model::node_id node_id) {
 
 ss::future<> brokers::apply(
-  const chunked_vector<metadata_response::broker>& brokers_metadata) {
+  const chunked_vector<describe_cluster_broker>& brokers_metadata) {
```
Is it intentional that DescribeCluster is now required, rather than fallback to Metadata?

michael-redpanda replied:
I could do one of two things:

1. This, or
2. Send 2 Metadata requests, one at version v10 (or 9) that has `include_cluster_authorized_operations`, and then one at v12 to get the topic IDs

Regardless, I will need to issue two calls to get the cluster authz ops and the topic IDs.

mmaslankaprv replied:
agree with Ben

BenPope commented on src/v/kafka/client/cluster.cc:
```diff
@@ -122,6 +122,13 @@ ss::future<api_version> get_metadata_request_version(
+ss::future<api_version> get_describe_cluster_request_version(
+  shared_broker_t broker,
+  std::optional<std::reference_wrapper<ss::abort_source>> as) {
+    co_return co_await get_required_api_version<describe_cluster_api>(
```
nitpick:
```suggestion
    return get_required_api_version<describe_cluster_api>(
```

michael-redpanda replied:
doh!

BenPope commented on src/v/kafka/client/cluster.cc:
```diff
@@ -137,45 +144,26 @@ ss::future<> cluster::initialize_metadata_with_seed() {
-         * the broker is not left in a connected state. This is important
-         * because the broker is created with unknown_node_id and we don't
-         * want to keep it connected.
-         */
```
This feels like a useful comment.

michael-redpanda replied:
it is!  adding it back

BenPope replied:
Could they be unremoved in the commit, rather than re-added later?

BenPope commented on src/v/kafka/client/cluster.cc:
```diff
@@ -216,38 +227,53 @@ cluster::dispatch_metadata_request(topic_request_list_t topic_request_list) {
+    vlog(_logger.debug, "Dispatching describe cluster request");
+
+    // Initialize the broker if it is not connected, this is required to
+    // make sure the broker has up to date API version information.
```
Where does this occur?

BenPope commented on src/v/kafka/client/cluster.cc:
```diff
@@ -216,38 +227,53 @@ cluster::dispatch_metadata_request(topic_request_list_t topic_request_list) {
-          "Metadata response is required to be returned as a result of "
-          "metadata request");
+    // Initialize the broker if it is not connected, this is required to
+    // make sure the broker has up to date API version information.
```
Where does this occur? Is it just part of `get_*_request_version`? I'd be tempted to delete the comment.

michael-redpanda replied:
All part of the call to get the API versions supported by the broker.  This call ends up calling the `get_supported_versions` method on `broker` (https://github.com/michael-redpanda/redpanda/blob/461577856e9b985d9a593686da31c0b5cdf5a62e/src/v/kafka/client/broker.cc#L345-L357).  This will connect the broker, perform authn, and then issue the APIVersions request.

BenPope commented on src/v/kafka/client/cluster.cc:
```diff
@@ -186,26 +174,49 @@ ss::future<> cluster::initialize_metadata_with_seed() {
+    try {
+        broker = _brokers.any();
+
+        auto cluster_describe_resp = co_await dispatch_describe_cluster_request(
+          broker, include_cluster_authz_t::yes);
+
+        if (!topic_request_list.has_value() || !topic_request_list->empty()) {
+            auto resp = co_await dispatch_metadata_request(
+              broker, std::move(topic_request_list));
+            metadata_resp = std::move(resp.data);
```
Could it make sense to issue a single metadata v10 if cluster authorized operations are required, and some (or all) topics are requested?

Should the request be made by topic id?

michael-redpanda replied:
We'd need to adjust how we handle the internal topic cache - right now we take the whole metadata request and apply it and rebuild the cache from the request.

> Should the request be made by topic id?

I think that's a good optimization to work in in the future.  Right now everything in cluster linking is done by name.

michael-redpanda replied:
Having the topic IDs returned in the metadata request is helpful in determining if the remote cluster has removed and re-created a topic of the same name.

BenPope replied:
v10 returns topic IDs and cluster auth ops.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/72247-->
<!--bd981559-8c1d-47e1-af7c-0792d9b299db-->
<!--2e8ae267-4b6d-452e-99aa-53fc0f35e003-->
<!--bd996bd6-d70f-478e-8134-81a9f0265beb-->
<!--707720ce-98f0-4872-a474-a8c19b6674a9-->
<!--7482b8df-a3f5-42ee-b19e-36c529a363f5-->
<!--title-->
## Retry command for [Build#72247](https://buildkite.com/redpanda/redpanda/builds/72247)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"sliding_window","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"adjacent_merge","enable_failures":false,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"adjacent_merge","enable_failures":false,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"adjacent_merge","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"chunked_sliding_window","enable_failures":false,"mixed_versions":true,"with_iceberg":false}
```
<!--/body-->

mmaslankaprv commented on src/v/kafka/client/cluster.cc:
```diff
@@ -137,45 +144,36 @@ ss::future<> cluster::initialize_metadata_with_seed() {
-          get_metadata_request_version(broker, _as));
-        if (request_version_f.failed()) {
-            auto ex = request_version_f.get_exception();
+        auto describe_cluster_f = co_await ss::coroutine::as_future(
+          dispatch_describe_cluster_request(
+            broker, include_cluster_authz_t::yes));
+
+        if (describe_cluster_f.failed()) {
+            auto ex = describe_cluster_f.get_exception();
```
With this change we no longer support clusters with earlier versions of Redpanda/Kafka. Since client implementation should remain version independent this logic is incorrect.

michael-redpanda replied:
fair - i'll adjust

michael-redpanda replied:
Putting RP aside, DescrbeClusters has been in Kafka since 2.8 so it's a pretty old API.

mmaslankaprv commented on src/v/kafka/client/cluster.cc:
```diff
@@ -124,45 +144,36 @@ ss::future<> cluster::initialize_metadata_with_seed() {
-          address.port(),
-          ex);
-        error = ex;
+        co_await apply_metadata(
```
Bot is correct here, previously this code was only called if there was no error, not that logic was removed

michael-redpanda replied:
No, if `describe_cluster_f.failed()` is true, that's caught in the above conditional

mmaslankaprv commented on src/v/kafka/client/cluster.cc:
```diff
@@ -173,46 +184,106 @@ ss::future<> cluster::initialize_metadata_with_seed() {
-          "metadata request");
+        broker = _brokers.any();
+
+        auto cluster_describe_resp = co_await dispatch_describe_cluster_request(
```
Again, here we lost support for cluster pre describe_cluster_request

mmaslankaprv replied:
This should be done conditionally, if describe_cluster is not supported we should fallback to metadata request

mmaslankaprv commented on src/v/kafka/client/brokers.h:
```diff
@@ -61,8 +61,7 @@ class brokers {
      */
-    ss::future<>
-    apply(const chunked_vector<metadata_response::broker>& brokers);
+    ss::future<> apply(const chunked_vector<describe_cluster_broker>& brokers);
```
we should keep the previous api in case the broker doesn't support describe cluster

mmaslankaprv commented:
It would be great to add a test for the cluster which doesn't support describe cluster API

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/72256-->
<!--091ee841-8bd7-4b2e-a6a1-171ce81c2d77-->
<!--79d355c9-8d49-40a0-b839-ea4d21c04d71-->
<!--8a69f6d2-56c3-4074-881d-0de65c6e0452-->
<!--title-->
## Retry command for [Build#72256](https://buildkite.com/redpanda/redpanda/builds/72256)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"sliding_window","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"sliding_window","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"sliding_window","enable_failures":false,"mixed_versions":true,"with_iceberg":false}
```
<!--/body-->

michael-redpanda committed:
k/c: Introduce metadata_update struct

This structure provides a common container for publishing metadata
updates when the cluster performs one.  This is helpful in future
commits where we introduce support for DescribeCluster which contains
much of the same information as Metadata but in a different format.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/c: Add ability to provide topic list when updating metadata

Callers to the request_metadata_update() method may now provide a list
of topics they wish to update.  If that list is std::nullopt, then all
topics will be requested.  An empty list requlests in no topics being
requested.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/c: Add ability to suggest version in metadata request

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/c: Rework dispatching and updating metadata

Separated out updating metadata from the dispatching of the metadata
request.  Will be useful in future commit where we start using the
DescribeCluster API.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/c: Wire in support for getting cluster authorized ops

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/client: Add support for DescribeCluster API

The DescribeCluster API carries the same information as the Metadata
request, however starting with Metadata v11, the cluster level
authorized operations were deprecated and moved to the new API. We still
want to use the most recent Metadata to get the topic IDs.  So this
update adds support for using the DescribeCluster API, if present, or
use Metadata v10 to get the cluster authorized ops.  Then a second
Metadata request is sent with the topics using the most up-to-date API.

As Redpanda currently supports DescribeCluster, we are able to test both
brokers with and without this API by simply not updating the
cluster_mock to support the DescribeCluster request.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
> It would be great to add a test for the cluster which doesn't support describe cluster API

See message in final commit, but since RP already supports DescribeClusters, the fixture tests cover testing the DescribeCluster API and we can test what happens with a cluster that doesn't support DC by simply not updating the cluster mock tests.

michael-redpanda commented:
Force push:

* Fixed issue causing failure in audit log test

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/72308-->
<!--bdd9ece4-6027-4d98-89fc-206918f819a9-->
<!--a471736b-f5d1-4cf6-afff-12ab2e0d5dbd-->
<!--6c4af79d-3c3a-4c91-ab25-70b7947b9644-->
<!--9dfbb0da-0c39-4583-b7f9-b7cf315f6564-->
<!--32349bd6-e6a4-49ae-bcd6-177b35c69dd1-->
<!--48617dfe-6861-429e-aef8-5be0b4a905f3-->
<!--title-->
## Retry command for [Build#72308](https://buildkite.com/redpanda/redpanda/builds/72308)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"adjacent_merge","enable_failures":false,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"chunked_sliding_window","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"sliding_window","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"sliding_window","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"adjacent_merge","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"adjacent_merge","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
```
<!--/body-->

mmaslankaprv commented on src/v/kafka/client/cluster.cc:
```diff
@@ -222,22 +369,27 @@ ss::future<> cluster::apply_metadata(metadata_response reply) {
-        _controller_id = reply.data.controller_id;
+        _controller_id = reply.controller_id;
+    }
+    if (!reply.topics.empty()) {
```
this logic seems to be incorrect, what if all topics in the cluster were simply removed ?

mmaslankaprv commented on src/v/kafka/client/types.h:
```diff
@@ -30,4 +32,42 @@ inline constexpr model::node_id consumer_replica_id{-1};
+        int32_t port{};
+        std::optional<ss::sstring> rack{};
+
+        friend std::ostream& operator<<(std::ostream& os, const broker& b);
```
please use `format_to` method instead of defining bot the stream operator and formatter

michael-redpanda replied:
Not sure if it's our version of gtest or fmtlib, but if you want to be able to print the structure in gtest, you have to define the stream operator

mmaslankaprv commented on src/v/kafka/client/types.h:
```diff
@@ -30,4 +32,42 @@ inline constexpr model::node_id consumer_replica_id{-1};
+/// \brief Structure used to report metadata updates
+/// Both describe cluster and metadata use very similar fields but are typed
+/// differently
+struct metadata_update {
```
This type doesn't differentiate between when topics were not requested vs they are not present

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/72331-->
<!--88407e94-abdc-4ac8-89d9-9e7048403244-->
<!--8b94be76-cd3d-4d77-a332-685aa86b4c8a-->
<!--dc8c0fee-4c47-48a5-9151-b03a20f475ac-->
<!--f6467d36-7ded-4f5c-9643-9b640ed68e0b-->
<!--f7d1d312-ab12-45cd-aa94-b0f97448b83c-->
<!--a5eef38d-65e1-4d64-a5bf-9b56954a0680-->
<!--bf2860d0-cbc3-476d-a1df-7fd0867e0843-->
<!--title-->
## Retry command for [Build#72331](https://buildkite.com/redpanda/redpanda/builds/72331)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"sliding_window","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"sliding_window","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"chunked_sliding_window","enable_failures":false,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"sliding_window","enable_failures":false,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"sliding_window","enable_failures":false,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"chunked_sliding_window","enable_failures":false,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"chunked_sliding_window","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/72406-->
<!--22f6f49f-4037-45e9-b027-76b7c5f04e24-->
<!--ac9b1d42-29f8-4c15-9d5a-ff72837df029-->
<!--d7fc87ff-1cc7-4135-bcb9-741708a4816f-->
<!--598aab70-4a7e-45dd-a802-674cb54853d1-->
<!--3af35c2c-d076-439f-a4dd-19769123dd6d-->
<!--title-->
## Retry command for [Build#72406](https://buildkite.com/redpanda/redpanda/builds/72406)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"adjacent_merge","enable_failures":false,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"sliding_window","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"sliding_window","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"adjacent_merge","enable_failures":false,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"chunked_sliding_window","enable_failures":false,"mixed_versions":true,"with_iceberg":false}
```
<!--/body-->

mmaslankaprv commented:
looks great 🔥

BenPope commented on src/v/kafka/client/cluster.cc:
```diff
@@ -90,24 +90,35 @@ ss::future<> cluster::request_metadata_update() {
-    // where NULL topic list represents request to list all topics)
-    if (!supported_version || supported_version->max < api_version(1)) {
+
+template<typename T>
```
Nitpick:
```suggestion
template<KafkaApi Api>
```
(And associated followups)

michael-redpanda commented:
Force push:

* Rebased off of dev to see if that fixed the weird errors with the random_node_operations_test

BenPope commented on src/v/kafka/client/cluster.cc:
```diff
@@ -18,6 +18,30 @@
+        update.brokers.push_back(
+          metadata_update::broker{
+            .node_id = b.node_id,
+            .host = b.host,
```
nitpick, since rack is moved
```suggestion
            .host = std::move(b.host),

BenPope commented on src/v/kafka/client/cluster.cc:
```diff
@@ -18,6 +18,30 @@
+            .port = b.port,
+            .rack = std::move(b.rack)});
+    }
+    update.cluster_id = r.data.cluster_id;
```
move?

BenPope commented on src/v/kafka/client/cluster.cc:
```diff
@@ -96,21 +96,23 @@ ss::future<> cluster::stop() {
 
-ss::future<> cluster::update_metadata() {
+ss::future<> cluster::update_metadata(
+  std::optional<chunked_vector<model::topic>> topics_request_list) {
```
Is it worth adding a using declaration `metadata_update::topics`?

michael-redpanda replied:
Let me get back to this when we implement topic ID support, because then i think it makes sense?

BenPope commented on src/v/kafka/client/cluster.cc:
```diff
@@ -242,13 +243,17 @@ ss::future<> cluster::dispatch_metadata_request(
+        auto metadata_version
+          = requested_version.has_value()
+              ? std::min(requested_version.value(), request_version)
+              : request_version;
```
nitpick:
```suggestion
        auto metadata_version = std::min(
          requested_version.value_or(request_version), request_version);
```

BenPope commented on src/v/kafka/client/cluster.cc:
```diff
@@ -236,35 +254,24 @@ ss::future<> cluster::dispatch_metadata_request(
-          _logger.warn, "Failed to dispatch metadata request - {}", e.what());
-    }
+    auto request_version = co_await get_metadata_request_version(broker, _as);
+    auto metadata_version = requested_version.has_value()
+                              ? std::min(
+                                  requested_version.value(), request_version)
+                              : request_version;
```
I wonder if it's worth passing the `requested_version` into `get_metadata_request_version` if this is going to be a common pattern for other requests.

nitpick:
```suggestion
    auto metadata_version = std::min(
      requested_version.value_or(request_version), request_version);
```

BenPope commented on src/v/kafka/client/cluster.cc:
```diff
@@ -40,6 +40,22 @@ metadata_update to_metadata_update(metadata_response r, bool topics_requested) {
+        update.brokers.push_back(
+          metadata_update::broker{
+            .node_id = b.broker_id,
+            .host = b.host,
```
move?

BenPope commented on src/v/kafka/client/cluster.cc:
```diff
@@ -40,6 +40,22 @@ metadata_update to_metadata_update(metadata_response r, bool topics_requested) {
+            .port = b.port,
+            .rack = std::move(b.rack)});
+    }
+    update.cluster_id = r.data.cluster_id;
```
move?

BenPope commented on src/v/kafka/client/cluster.cc:
```diff
@@ -122,9 +138,31 @@ ss::future<> cluster::dispatch_and_apply_metadata_updates(
+            // cluster level authorized operations
+            auto metadata_resp = co_await dispatch_metadata_request(
+              broker, {}, api_version(10));
+            mu = to_metadata_update(std::move(metadata_resp), false);
```
It's not obvious what this `false` is for. Could use named_bool or named the variable.

BenPope commented on src/v/kafka/client/cluster.cc:
```diff
@@ -122,9 +138,31 @@ ss::future<> cluster::dispatch_and_apply_metadata_updates(
+        if (!topics_request_list.has_value() || !topics_request_list->empty()) {
+            // If there are topics to request, then do another metadata request
+            // with the most up-to-date version
+            auto metadata_resp = co_await dispatch_metadata_request(
```
Could this be combined with the previous request to metadata?

michael-redpanda replied:
No - the first one is v10 so I can get the cluster level authz, the second one will use the most recent one so we can get topic ids

BenPope replied:
> No - the first one is v10 so I can get the cluster level authz, the second one will use the most recent one so we can get topic ids

Metadata v10 responds with topic ids, although they cannot be requested, at least for Redpanda v25.2.x and the Kafka 4.0.0.

BenPope commented on src/v/kafka/client/cluster.cc:
```diff
@@ -181,49 +226,50 @@ ss::future<> cluster::initialize_metadata_with_seed() {
+        auto describe_cluster_version_f = co_await ss::coroutine::as_future(
+          broker->get_supported_versions(describe_cluster_api::key, _as));
+        if (describe_cluster_version_f.failed()) {
+            auto ex = describe_cluster_version_f.get_exception();
```
Looks like this could assign to error directly.

michael-redpanda commented:
Force push:

* Addressed PR comments

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/72443-->
<!--fe415399-b8c4-475f-9a87-d1352faa9448-->
<!--f4eb59cb-ed5b-4705-afd5-8c57a2e23c8c-->
<!--f5f86899-20b5-461e-863d-f79de4759a26-->
<!--title-->
## Retry command for [Build#72443](https://buildkite.com/redpanda/redpanda/builds/72443)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"sliding_window","enable_failures":false,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"adjacent_merge","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"chunked_sliding_window","enable_failures":false,"mixed_versions":true,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"adjacent_merge","enable_failures":true,"mixed_versions":true,"with_iceberg":false}
```
<!--/body-->

michael-redpanda commented:
I think I need this: https://github.com/redpanda-data/redpanda/pull/27618

Merged and backported (and maybe released?) before the random_node_operations test starts passing

michael-redpanda commented:
Force push:

* Rebased off of dev to fix conflict
